src/hotspot/share/runtime/os.cpp

```diff
@@ -725,11 +701,19 @@ void* os::realloc(void *memblock, size_t size, MEMFLAGS memflags, const NativeCa
       return NULL;
     }
 
+    const size_t old_size = MallocTracker::malloc_header(memblock)->size();
+
+    // Observe MallocLimit
+    if (size > old_size && MemTracker::check_exceeds_limit(size - old_size, memflags)) {
```

Comment:
Can we make it a bit easier to parse by rewriting it as:

`    if ((size > old_size) && MemTracker::check_exceeds_limit(size - old_size, memflags)) {`

To me personally that is easier to read.

I see a bigger issue here, however. In the worse case scenario where the os is unable to expand the memory chunk in place, it will need to allocate `size` bytes in a new region, so the total **potential** resources from the os point of view is `size+old_size` bytes. Shouldn't we assume this worse case and test for that?

I'd rather get some false positives that miss a single true positive... 

And if that were the case we wouldn't need to check for `(size > old_size) ` and we should have just:

`    if (MemTracker::check_exceeds_limit(size + old_size, memflags)) {`

with a comment explaining this choice.

Reply:
> Can we make it a bit easier to parse by rewriting it as:
> 
> ` if ((size > old_size) && MemTracker::check_exceeds_limit(size - old_size, memflags)) {`
> 
> To me personally that is easier to read.
> 

Sure

> I see a bigger issue here, however. In the worst case scenario, where the os is unable to expand the memory chunk in place, it will need to allocate `size` bytes in a new region, so the total **potential** resources from the os point of view is `size+old_size` bytes for that particular allocation. Shouldn't we assume this worst case and test for that? I.e.
> 
> ` if ((size > old_size) && MemTracker::check_exceeds_limit(size, memflags)) {`
> 
> We would need a comment here explaining this choice if we make this change.
> 
> I'd rather get some false positives that miss a single true positive...

Nah, no need to overthink this. 

If you try to predict how much memory your malloc causes your process to allocate, it gets very fuzzy very quickly:

Below are at least two allocating layers (libc and the kernel's vm manager). Both of them balance memory and CPU footprint and allocation speed. No libc or kernel is the same, too. Each layer below you may return cached or partly cached memory and apply an often sizable overhead to what you are allocating. So our mallocs and frees have only a very indirect effect on RSS. And we don't even see all mallocs here, just the ones from libjvm.

Therefore it is best to limit the meaning of MallocLimit to "limit to how much hotspot is allowed to allocate". That is also the most useful. For limiting the memory of a process, different os-side tools exist (cgroup limits, ulimit, etc).

Reply:
The entire point of this effort is to catch elusive OOM errors. You even moved the check for the memory requirements before we actually request it, from the current way where we check the limit only after we acquire new memory.

Isn't assuming the worst case scenario, where realloc(), might allocate new memory first, in the same vein as that and we are just being cautious, to catch those hard to reproduce bugs?

This cases might not occur very often, granted, but why not catch them if we can?

Aren't there any platforms where we run Java where realloc() is implemented in terms of malloc() and memcpy()?

Reply:
> The entire point of this effort is to catch elusive OOM errors. You even moved the check for the memory limits before we actually allocate the memory, from the current way where we check the limit only after we acquire new memory.
> 
> Isn't assuming the worst case scenario, where realloc(), might allocate new memory first, in the same vein as that and we are just being cautious, to catch those hard to reproduce bugs?
> 
> Basically we are saying that if the OS had to allocate new memory on top of the old one and doing so would exceed the limit we would like to know about it. Will it happen often? No? Will it happen during the lifetime of a memory intensive process? Most likely yes.
> 
> These cases might not occur very often, granted, but why not catch them if we can?
> 
> Aren't there any platforms where we run Java where realloc() is implemented in terms of malloc() and memcpy()?

I think its just wrong. False positives are not helpful.

1) its semantically wrong. The malloc limit shall triggers when our global or category-specific malloc load exceeds a given threshold. Be it either the malloc that just happened (old version) or the malloc that we are about to do (new version). But I don't want the mechanism to second-guess what I'm telling it and proactively trigger alerts. Example: I set 10m as limit and have a 5MB buffer. Expanding this buffer to 5.5 MB should not trigger any alert. I don't want false positives, they are not helpful.

2) You try to predict an OOM kill caused by a malloc event, right? For that to work the prediction must be at least somewhat correct. But you only have very limited knowledge. Some examples:
- The allocation request is fulfilled from cached memory. No new memory needed at all. The smaller the allocation the more likely this is, and if the libc uses sbrk exclusively, this is very likely.
- The libc always has overhead per allocation. E.g. glibc, at the very least, ~40ish bytes. Should I account for this? With many fine-grained allocations this can have an enourmous impact. But overhead depends on a number of factors, and every libc is different.
- The allocation request is large and the libc decides to mmap it (not all do). So we allocate in page granularity. Should I now adjust my prediction for page size? What page size? The libc may still reuse already mapped memory though. Or the underlying kernel vm may do that.
- The realloc is done by a new thread, and the libc uses thread-specific memory pools. A new memory pool is created. Is this memory committed immediately (e.g. older glibcs) or committed on demand? Depending on the answer, we may see a large bump in RSS or none at all. Depends on libc and libc version.
- The realloc is large and cannot be enlarged in place. But I could think of several ways of enlarging such an allocation out-of-place without needing memory * 2. If even I can think of those, libc devs probably do too.

So, these examples show that an allocation may not move the needle at all or may move the needle much farther than you think. Trying to guess this just makes the code complex and causes false positives that are difficult to argue about. All without having any real benefit.

Reply:
Thank you for taking the time to explain.

I'm worrying about those hard to pin mysterious issues that we have plenty of that possibly could be related to memory pressure, but I understand your point of view and will defer to you (even if there is a 5% of this that continues to bother me).

Reply:
I understand. This is not the tool to solve all that, but it can help. 

There are other tools. At SAP, we added something called "High Memory Reports", the ability to automatically generate customizable reports when certain memory threshold is reached [1]. In addition to that, we have our SapMachine Vitals [2], which are awesome  - basically like a little inbuilt pidstat but with a lot more information. For instance, it tells you the history of c-heap allocations (both from the JVM view and from the libc's view).

Would love to see this in OpenJDK, but have zero hopes of doing so - tried to get the Vitals in years ago but hit a wall of resistance from JFR people.

[1] https://github.com/SAP/SapMachine/wiki/SapMachine-High-Memory-Reports
[2] https://github.com/SAP/SapMachine/wiki/SapMachine-Vitals